sklearn/ensemble/iforest.py

```diff
@@ -151,15 +151,27 @@ def fit(self, X, y=None, sample_weight=None):
         y = rnd.uniform(size=X.shape[0])
 
         # ensure that max_sample is in [1, n_samples]:
-        max_samples = self.max_samples
         n_samples = X.shape[0]
-        if max_samples > n_samples:
-            warn("max_samples (%s) is greater than the "
-                 "total number of samples (%s). max_samples "
-                 "will be set to n_samples for estimation."
-                 % (self.max_samples, n_samples))
-            max_samples = n_samples
 
+        if self.max_samples == 'auto':
+            max_samples = min(256, n_samples)
+        else:
+            if isinstance(self.max_samples, (numbers.Integral, np.integer)):
+                if self.max_samples > n_samples:
+                    warn("max_samples (%s) is greater than the "
+                         "total number of samples (%s). max_samples "
+                         "will be set to n_samples for estimation."
+                         % (self.max_samples, n_samples))
+                    max_samples = n_samples
+                else:
+                    max_samples = self.max_samples
+            else: # float
+                if not (0 < self.max_samples <= 1.):
+                    raise ValueError("max_samples must be in (0, 1]")
+                max_samples = int(self.max_samples * X.shape[0])
+
+        self.max_samples_ = max_samples
+        self.base_estimator.max_depth = int(np.ceil(np.log2(max(max_samples, 2))))
```

Comment:
It must be on base_estimator_, has you cannot modify an init argument at any time.